iroh-net/src/magicsock.rs

```diff
@@ -278,6 +286,664 @@ impl Inner {
+
+        if self.is_closed() {
+            inc_by!(MagicsockMetrics, send_data_network_down, bytes_total as _);
+            return Poll::Ready(Err(io::Error::new(
+                io::ErrorKind::NotConnected,
+                "connection closed",
+            )));
+        }
+
+        let mut n = 0;
+        if transmits.is_empty() {
+            return Poll::Ready(Ok(n));
+        }
+
+        trace!(
+            "sending:\n{}",
+            transmits.iter().fold(
+                String::with_capacity(transmits.len() * 50),
+                |mut final_repr, t| {
+                    final_repr.push_str(
+                        format!(
+                            "  dest: {}, src: {:?}, content_len: {}\n",
+                            QuicMappedAddr(t.destination),
+                            t.src_ip,
+                            t.contents.len()
+                        )
+                        .as_str(),
+                    );
+                    final_repr
+                }
+            )
+        );
+
+        let dest = transmits[0].destination;
+        for transmit in transmits.iter() {
+            if transmit.destination != dest {
+                break;
+            }
+            n += 1;
+        }
+        // TODO: This is the remaining alloc on the hot path for send.
+        // Unfortunately I don't see a way around this because we have do modify the transmits.
+        let mut transmits = transmits[..n].to_vec();
+        let dest = QuicMappedAddr(dest);
+
+        match self.peer_map.get_send_addrs_for_quic_mapped_addr(&dest) {
+            // TODO: Handle ping actions
+            Some((public_key, udp_addr, derp_region, mut msgs)) => {
+                trace!(peer = %public_key.fmt_short(), quic_addr = %dest, n = %transmits.len(), "send");
+                // If we have pings to send, we *have* to send them out first.
+                if !msgs.is_empty() {
+                    if let Err(err) = ready!(self.poll_handle_ping_actions(cx, &mut msgs)) {
+                        warn!(peer = %public_key.fmt_short(), "failed to handle ping actions: {err:?}");
+                    }
+                }
+
+                let mut udp_sent = false;
+                let mut derp_sent = false;
+                let mut udp_error = None;
+
+                // send udp
+                if let Some(addr) = udp_addr {
+                    // rewrite target addresses.
+                    for t in transmits.iter_mut() {
+                        t.destination = addr;
+                    }
+                    match ready!(self.poll_send_udp(addr, &transmits, cx)) {
+                        Ok(n) => {
+                            debug!(peer = %public_key.fmt_short(), ?addr, ?n, "sent udp");
+                            // truncate the transmits to be sent over derp, otherwise we'd
+                            // potentially send the same transmits multiple times over derp.
```

Comment:
the system relies on double sending over both udp and derp, if there are both addresses available, so this needs to be removed

Reply:
To clarify my thought: The double sending should still be working fine with this change. If `poll_send_udp`returns `Poll::Ready(n)` we'd only send those `n` transmits to Derp, but we return this `n` from `poll_send` so on the next call (which happens immediately) we'll get passed the next transmits after `n`. Procedure repeats. So if my understanding is correct without the truncate we'd double-send to Derp the transmits following `n`(because they'll be sent on this and the next call to `poll_send` from quinn). Which might be just fine though.

Reply:
I see, okay, then we should do this, makes sense 👍

Reply:
Will restore!